File: viewer/src/main/java/org/cryptobiotic/rlauxe/viewer/BeanProperties.java
package org.cryptobiotic.rlauxe.viewer;

import org.cryptobiotic.rlauxe.audit.ContestRound;
import org.cryptobiotic.rlauxe.audit.CountyPools;
import org.cryptobiotic.rlauxe.core.Assertion;
import org.cryptobiotic.rlauxe.core.ContestWithAssertions;
import ucar.ui.prefs.BeanTable;

import java.io.FileOutputStream;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class BeanProperties {
    static ArrayList<BeanTable.TableBeanProperty> contests = new ArrayList<>();
    static {
        contests.add(new BeanTable.TableBeanProperty("id", "contest identifier"));
        contests.add(new BeanTable.TableBeanProperty("name", "contest name"));
        contests.add(new BeanTable.TableBeanProperty("type", "contest type"));
        contests.add(new BeanTable.TableBeanProperty("NCand", "number of candidates"));
        contests.add(new BeanTable.TableBeanProperty("winners", "list of winning candidates"));
        contests.add(new BeanTable.TableBeanProperty("nc", "trusted upper bound on contest number of cards"));
        contests.add(new BeanTable.TableBeanProperty("npop", "population size (for diluted margin)"));
        contests.add(new BeanTable.TableBeanProperty("ncast", "number of cards cast"));
        contests.add(new BeanTable.TableBeanProperty("population", "Nc (hasStyle) or Npop (noStyle)"));
        contests.add(new BeanTable.TableBeanProperty("phantoms", "number of phantom cards"));
        contests.add(new BeanTable.TableBeanProperty("status", "status of contest"));
        contests.add(new BeanTable.TableBeanProperty("votes", "reported vote count"));
        contests.add(new BeanTable.TableBeanProperty("undervotes", "reported undervote count"));
        contests.add(new BeanTable.TableBeanProperty("uvPct", "percent undervote count"));

        contests.add(new BeanTable.TableBeanProperty("voteDiff", "(winner-loser) votes (smallest assertion)"));
        contests.add(new BeanTable.TableBeanProperty("margin", "voteDiff / population size (smallest assertion)"));
        contests.add(new BeanTable.TableBeanProperty("dilutedMargin", "voteDiff/Npop (smallest assertion)"));
        contests.add(new BeanTable.TableBeanProperty("reportedMargin", "voteDiff/Nc (smallest assertion)"));
        contests.add(new BeanTable.TableBeanProperty("recountMargin", "(winner-loser)/winner (smallest assertion)"));
        contests.add(new BeanTable.TableBeanProperty("noerror", "assort value when mvr agrees with the cvr"));
        contests.add(new BeanTable.TableBeanProperty("payoff", "payoff factor for each mvr that agrees with the cvr"));

        contests.add(new BeanTable.TableBeanProperty("estMvrs", "estimated number of mvrs needed"));
        contests.add(new BeanTable.TableBeanProperty("estRisk", "estimated risk"));
        contests.add(new BeanTable.TableBeanProperty("haveMvrs", "number of mvrs that were sampled for this contest"));
        contests.add(new BeanTable.TableBeanProperty("mvrsExtra", "(haveMvrs-estMvrs)"));
        contests.add(new BeanTable.TableBeanProperty("mvrsUsed", "number of mvrs actually used during audit"));
        contests.add(new BeanTable.TableBeanProperty("samplePct", "estMvrs/population size"));

        // Corla
        contests.add(new BeanTable.TableBeanProperty("target", "is a Corla targeted contest"));
        contests.add(new BeanTable.TableBeanProperty("NCounties", "number of counties, or the county name if only one"));
        contests.add(new BeanTable.TableBeanProperty("corlaEst", "estimate from Corla (super simple) algorithm"));
        contests.add(new BeanTable.TableBeanProperty("nvotes", "tabulated nvotes from auditcenter or cvrs"));
        contests.add(new BeanTable.TableBeanProperty("compareNvotes", "tabulated nvotes from auditcenter - cvrs"));
        contests.add(new BeanTable.TableBeanProperty("pctCompareNvotes", "compareNvotes / (auditcenter nvotes)"));

        // CountyPoolsTable.Contest
        contests.add(new BeanTable.TableBeanProperty("contestId", "contest identifier"));
        contests.add(new BeanTable.TableBeanProperty("contestName", "contest name"));
        contests.add(new BeanTable.TableBeanProperty("countyPopulation", "contestRound.ballotCardCount"));
        contests.add(new BeanTable.TableBeanProperty("contestPopulation", "contestRound.contestBallotCardCount"));
        contests.add(new BeanTable.TableBeanProperty("corlaSample", "contestRound.estimatedSamplesToAudit"));
        contests.add(new BeanTable.TableBeanProperty("corlaHaveMvrs", "estimated Corla uniform have mvrs"));
        contests.add(new BeanTable.TableBeanProperty("corlaRisk", "estimated Corla uniform risk"));
        contests.add(new BeanTable.TableBeanProperty("cvrNcards", "number or cards from cvrs"));
        contests.add(new BeanTable.TableBeanProperty("diffNcards", "contest.Nc  - cvrNcards"));
        contests.add(new BeanTable.TableBeanProperty("pctMissing", "diffNcards / contest.Nc"));
        contests.add(new BeanTable.TableBeanProperty("diffNvotes", "votes.sum() - cvrNvotes"));

        // AuditRound
        contests.add(new BeanTable.TableBeanProperty("round", "index of audit round"));
        contests.add(new BeanTable.TableBeanProperty("estNewMvrs", "estimated new samples needed"));
        contests.add(new BeanTable.TableBeanProperty("estPct", "estimated samples needed / contest Nc"));
        contests.add(new BeanTable.TableBeanProperty("haveNewMvrs", "new contest cards in sample"));
        contests.add(new BeanTable.TableBeanProperty("mvrLimit", "limit on number of mvrs to audit; set by auditor"));
        contests.add(new BeanTable.TableBeanProperty("risk", "measured risk"));

        contests.add(new BeanTable.TableBeanProperty("include", "include the contest in this audit round"));
        contests.add(new BeanTable.TableBeanProperty("done", "contest has completed"));
    }

    static ArrayList<BeanTable.TableBeanProperty> assertions = new ArrayList<>();
    static {
        assertions.add(new BeanTable.TableBeanProperty("type", "assorter type"));
        assertions.add(new BeanTable.TableBeanProperty("winner", "assertion winner candidate"));
        assertions.add(new BeanTable.TableBeanProperty("loser", "assertion loser candidate"));
        assertions.add(new BeanTable.TableBeanProperty("desc", "assertion description"));
        assertions.add(new BeanTable.TableBeanProperty("difficulty", "assertion difficulty measure"));
        assertions.add(new BeanTable.TableBeanProperty("name", "assertion short name"));

        assertions.add(new BeanTable.TableBeanProperty("upper", "assorter upper bound"));
        assertions.add(new BeanTable.TableBeanProperty("margin", "voteDiff / population size"));
        assertions.add(new BeanTable.TableBeanProperty("noerror", "assort value when mvr agrees with the cvr"));
        assertions.add(new BeanTable.TableBeanProperty("payoff", "payoff factor for each mvr that agrees with the cvr"));
        assertions.add(new BeanTable.TableBeanProperty("estMvrs", "estimated number of mvrs needed"));
        assertions.add(new BeanTable.TableBeanProperty("estRisk", "estimated risk"));

        assertions.add(new BeanTable.TableBeanProperty("mean", "average assorter value"));
        assertions.add(new BeanTable.TableBeanProperty("recountMargin", "(winner-loser)/winner"));

        assertions.add(new BeanTable.TableBeanProperty("round", "index of audit round"));
        assertions.add(new BeanTable.TableBeanProperty("prevMvrs", "mvrs from previous rounds"));
        assertions.add(new BeanTable.TableBeanProperty("estNewMvrs", "estimated new samples needed"));
        assertions.add(new BeanTable.TableBeanProperty("mvrsUsed", "mvrs used in this round"));
        assertions.add(new BeanTable.TableBeanProperty("status", "status of assertion completion"));
        assertions.add(new BeanTable.TableBeanProperty("risk", "measured risk (minimum PValue of audit)"));
        assertions.add(new BeanTable.TableBeanProperty("completed", "round that assertions was proved"));
    }

    static ArrayList<BeanTable.TableBeanProperty> countyPools = new ArrayList<>();
    static {
        countyPools.add(new BeanTable.TableBeanProperty("countyName", "county name"));
        countyPools.add(new BeanTable.TableBeanProperty("countyPoolId", "county name"));
        countyPools.add(new BeanTable.TableBeanProperty("population", "county population from round.ballotCardCount"));
        countyPools.add(new BeanTable.TableBeanProperty("totalCards", "number of cvrs in the pool"));
        countyPools.add(new BeanTable.TableBeanProperty("diff", "population - totalCards"));
        countyPools.add(new BeanTable.TableBeanProperty("diffPct", "(population - totalCards)/population"));
        countyPools.add(new BeanTable.TableBeanProperty("diffNVotes", "auditcenter.votes - cvr.votes"));
        countyPools.add(new BeanTable.TableBeanProperty("pctDiffNVotes", "(auditcenter.votes - cvr.votes)/auditcenter.votes"));
    }

    static public <T> String showContestG(T bean, BeanTable<T>.TableBeanModel tableModel, ContestWithAssertions cua) {
        StringBuilder sb = new StringBuilder();
        sb.append("%n%s%n".formatted( tableModel.showBean(bean, BeanProperties.contests)));
        if (cua != null) sb.append("\n%s%n".formatted(cua.show()));
        return sb.toString();
    }

    static public <T> String showAssertionG(T bean, BeanTable<T>.TableBeanModel tableModel, ContestWithAssertions cua, Assertion assertion) {
        StringBuilder sb = new StringBuilder();
        sb.append("%n%s%n".formatted(tableModel.showBean(bean, BeanProperties.assertions)));
        sb.append(assertion.show());
        sb.append("\n   difficulty: %s".formatted(cua.getContest().showAssertionDifficulty(assertion.getAssorter())));
        return sb.toString();
    }

    static public <T> String showCountyPoolsG(T bean, BeanTable<T>.TableBeanModel tableModel, CountyPools pool) {
        StringBuilder sb = new StringBuilder();
        sb.append("%n%s%n".formatted(tableModel.showBean(bean, BeanProperties.countyPools)));
        return sb.toString();
    }

    static public <T> String printTableG(List<T> beans, BeanTable<T>.TableBeanModel tableModel, List<BeanTable.TableBeanProperty> properties, String name) {
        StringBuilder sb = new StringBuilder();
        StringBuilder sb2 = new StringBuilder();

        String header = tableModel.beanTableHeader(properties);
        sb.append(header);
        sb2.append("| " + header.replace(',','|'));

        for (var bean : beans) {
            String beanCsv = tableModel.beanCsv(bean, properties);
            sb.append(beanCsv);
            sb2.append("| " + beanCsv.replace(',','|'));
        }

        var file = "/home/stormy/rla/temp/"+name+".csv";
        try (FileOutputStream fout = new FileOutputStream(file);
             OutputStreamWriter writer = new OutputStreamWriter(fout, StandardCharsets.UTF_8)) {
            writer.write(sb.toString());
            writer.flush();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        return sb2.toString();
    }

}

File: viewer/src/main/kotlin/org/cryptobiotic/rlauxe/beans/BeanProperties.kt
package org.cryptobiotic.rlauxe.beans

import org.cryptobiotic.rlauxe.core.Assertion
import org.cryptobiotic.rlauxe.core.ContestWithAssertions
import java.io.FileOutputStream
import java.io.OutputStreamWriter
import java.nio.charset.StandardCharsets

object BeanProperties {
    fun getContestBeanProperties() = contests // for java access

    // used by AuditRoundsTable.contestRoundTable
    // ContestsPanel.contestTable
    // BelgiumContestsTable.contestTable
    // CorlaContestsTable.contestTable and SamplingTable.contestTable (same CorlaContestBean)
    var contests = mutableListOf<TableBeanProperty>()

    init {
        contests.add(TableBeanProperty("NCand", "number of candidates"))
        contests.add(TableBeanProperty("NCounties", "number of counties, or the county name if only one"))
        contests.add(TableBeanProperty("contestId", "contest identifier"))
        contests.add(TableBeanProperty("cvrNcards", "number of cards from cvrs"))
        contests.add(TableBeanProperty("diffNcards", "contest.Nc  - cvrNcards"))

        contests.add(TableBeanProperty("estMvrs", "estimated number of mvrs needed"))
        contests.add(TableBeanProperty("estRisk", "estimated risk"))
        contests.add(TableBeanProperty("haveMvrs", "number of mvrs that were sampled for this contest"))
        contests.add(TableBeanProperty("id", "contest identifier"))
        contests.add(TableBeanProperty("include", "include the contest in this audit round"))
        contests.add(TableBeanProperty("margin", "voteMargin / population size (smallest assertion)"))
        contests.add(TableBeanProperty("maxRisk", "when included, the maximum risk allowed"))
        // contests.add(TableBeanProperty("missing", "TODO"))
        contests.add(TableBeanProperty("mvrLimit", "limit on number of mvrs to audit; set by auditor"))
        contests.add(TableBeanProperty("mvrsExtra", "(haveMvrs-estMvrs)"))
        contests.add(TableBeanProperty("mvrsUsed", "number of mvrs actually used during audit"))
        contests.add(TableBeanProperty("name", "contest name"))
        contests.add(TableBeanProperty("nc", "trusted upper bound on contest number of cards"))
        // contests.add(TableBeanProperty("ncast", "number of cards cast"))
        contests.add(TableBeanProperty("noerror", "assort value when mvr agrees with the cvr"))
        contests.add(TableBeanProperty("npop", "Npop size (for diluted margin)"))
        contests.add(TableBeanProperty("nvotes", "tabulated nvotes from auditcenter or cvrs"))
        contests.add(TableBeanProperty("payoff", "payoff factor for each mvr that agrees with the cvr"))
        contests.add(TableBeanProperty("population", "Nc (hasStyle) or Npop (noStyle)"))
        contests.add(TableBeanProperty("phantoms", "number of phantom cards"))
        contests.add(TableBeanProperty("recountMargin", "(winner-loser)/winner (smallest assertion)"))
        contests.add(TableBeanProperty("risk", "measured risk"))
        contests.add(TableBeanProperty("source", "auditcenter or cvrs"))
        contests.add(TableBeanProperty("status", "status of contest"))
        contests.add(TableBeanProperty("target", "is a Corla targeted contest"))
        contests.add(TableBeanProperty("type", "contest type"))
        contests.add(TableBeanProperty("undervotes", "reported undervote count"))
        contests.add(TableBeanProperty("uvPct", "percent undervote count"))
        contests.add(TableBeanProperty("voteMargin", "(winner-loser) votes (smallest assertion)"))
        contests.add(TableBeanProperty("votes", "reported vote count"))
        contests.add(TableBeanProperty("winners", "list of winning candidates"))

        //contests.add(TableBeanProperty("dilutedMargin", "voteMargin/Npop (smallest assertion)"))
        //contests.add(TableBeanProperty("reportedMargin", "voteMargin/Nc (smallest assertion)"))
        // contests.add(TableBeanProperty("samplePct", "estMvrs/population size"))

        // Belgium
        contests.add(TableBeanProperty("NFailures", "number of contested assertions"))
        contests.add(TableBeanProperty("nseats", "nseats to win"))

        // ContestRound
        contests.add(TableBeanProperty("done", "contest has completed"))
        contests.add(TableBeanProperty("estNewMvrs", "estimated new samples needed"))
        contests.add(TableBeanProperty("estPct", "estimated samples needed / contest Nc"))
        contests.add(TableBeanProperty("haveNewMvrs", "new contest cards in sample"))
        contests.add(TableBeanProperty("maxIndex", "max sample index for this contest"))
        contests.add(TableBeanProperty("round", "index of audit round"))

        // Corla
        contests.add(TableBeanProperty("corlaHaveMvrs", "Corla mvrs for contest's strata (from contestComparisonFile)"))
        contests.add(TableBeanProperty("corlaRisk", "estimated Corla uniform risk"))
        contests.add(TableBeanProperty("corlaStrata", "Corla strata size (from contestRoundFile ballot_card_count)"))
        contests.add(TableBeanProperty("corlaVoteDiff", "TODO"))
        contests.add(TableBeanProperty("corlaVoteMargin", "Corla margin in votes (from contestRoundFile min_margin)"))
        contests.add(TableBeanProperty("cvrNvotes", "number of votes from cvrs"))
        contests.add(TableBeanProperty("cvrNundervotes", "number of undervotes from cvrs"))
        contests.add(TableBeanProperty("diffNvotes", "auditcenter.votes.sum() - cvrs.votes.sum()"))
        contests.add(TableBeanProperty("pctNcardsMissing", "diffNcards / contest.Nc"))
        contests.add(TableBeanProperty("pctNvotesMissing", "diffNvotes / auditcenter.votes.sum()"))
        contests.add(
            TableBeanProperty(
                "corlaEstMvrs",
                "Corla estimated nmvrs (from contestRoundFile estimated_samples_to_audit)"
            )
        )

        // CountyContest
        contests.add(TableBeanProperty("contestName", "contest name"))
        contests.add(
            TableBeanProperty(
                "compareNvotes",
                "tabulated nvotes from auditcenter - cvrs for this county and contest"
            )
        )
        contests.add(
            TableBeanProperty(
                "pctCompareNvotes",
                "compareNvotes / (auditcenter nvotes) for this county and contest"
            )
        )

        // ContestCounty
        contests.add(TableBeanProperty("countyName", "county name"))
        contests.add(TableBeanProperty("countyPoolId", "county pool id"))
        contests.add(TableBeanProperty("estNcards", "estimated ncards in this county and contest"))
        contests.add(TableBeanProperty("pctDiffNvotes", "diffNvotes / auditcenter.nvotes"))
        contests.add(TableBeanProperty("voteForN", "allowed number of votes"))
    }

    // CorlaContestsTable.contestCountyTable
    // CountyTable.countyContestTable
    // SamplingTable.CountyContest
    var assertions = mutableListOf<TableBeanProperty>()

    init {
        assertions.add(TableBeanProperty("desc", "assertion description"))
        assertions.add(TableBeanProperty("difficulty", "assertion difficulty measure"))
        assertions.add(TableBeanProperty("estMvrs", "estimated number of mvrs needed"))
        assertions.add(TableBeanProperty("estRisk", "estimated risk"))

        assertions.add(TableBeanProperty("loser", "assertion loser candidate"))
        assertions.add(TableBeanProperty("margin", "voteMargin / population size"))
        assertions.add(TableBeanProperty("mean", "average assorter value"))
        assertions.add(TableBeanProperty("name", "assertion short name"))
        assertions.add(TableBeanProperty("noerror", "assort value when mvr agrees with the cvr"))
        assertions.add(TableBeanProperty("payoff", "payoff factor for each mvr that agrees with the cvr"))

        assertions.add(TableBeanProperty("recountMargin", "(winner-loser)/winner"))
        assertions.add(TableBeanProperty("type", "assorter type"))
        assertions.add(TableBeanProperty("upper", "assorter upper bound"))
        assertions.add(TableBeanProperty("winner", "assertion winner candidate"))

        // assertion round
        assertions.add(TableBeanProperty("completed", "round that assertions was proved"))
        assertions.add(TableBeanProperty("estNewMvrs", "estimated new samples needed"))
        assertions.add(TableBeanProperty("mvrsUsed", "mvrs used in this round"))
        assertions.add(TableBeanProperty("prevMvrs", "mvrs from previous rounds"))
        assertions.add(TableBeanProperty("round", "index of audit round"))
        assertions.add(TableBeanProperty("status", "status of assertion completion"))
        assertions.add(TableBeanProperty("risk", "measured risk (minimum PValue of audit)"))

        // belgium dhondt
        assertions.add(TableBeanProperty("scoreDiff", "(winner-loser) score"))
        assertions.add(TableBeanProperty("scoreRange", "max scoreDiff for this number of samples"))
    }
}

fun <T> showContestWithDesc(bean: T, tableModel: BeanTableModel<T>, cua: ContestWithAssertions?) = buildString {
    appendLine(tableModel.showBean(bean, BeanProperties.contests))
    if (cua != null) append(cua.show())
}

fun <T> showAssertionWithDesc(bean: T, tableModel: BeanTableModel<T>, cua: ContestWithAssertions, assertion: Assertion) = buildString {
    appendLine(tableModel.showBean(bean, BeanProperties.assertions))
    appendLine(assertion.show())
    appendLine("difficulty: ${cua.contest.showAssertionDifficulty(assertion.assorter)}")
}

fun <T> printTable(
    beanTable: BeanTable<T>,
    properties: List<TableBeanProperty>,
    name: String,
): String {
    val indexBeans = beanTable.indexedBeans()
    val sortedIndexBeans = indexBeans.sortedBy { it.viewIndex }
    val sortedBeans = sortedIndexBeans.map { it.bean }

    val sb = StringBuilder()
    val sb2 = StringBuilder()

    val header: String = beanTable.tableModel.beanTableHeader(properties)
    sb.append(header)
    sb2.append("| " + header.replace(',', '|'))

    for (bean in sortedBeans) {
        val beanCsv: String = beanTable.tableModel.beanCsv(bean, properties)
        sb.append(beanCsv)
        sb2.append("| " + beanCsv.replace(',', '|'))
    }

    val file = "/home/stormy/rla/temp/" + name + ".csv"
    try {
        FileOutputStream(file).use { fout ->
            OutputStreamWriter(fout, StandardCharsets.UTF_8).use { writer ->
                writer.write(sb.toString())
                writer.flush()
            }
        }
    } catch (e: Exception) {
        throw RuntimeException(e)
    }

    return sb2.toString()
}
